storage/perfschema/table_malloc_stats.cc

```diff
@@ -0,0 +1,269 @@
+
+  for (; (f = *fields); fields++) {
+    if (read_all || bitmap_is_set(table->read_set, f->field_index())) {
+      DBUG_ASSERT(f->field_index() < NUM_SUMMARY_STAT);
+      set_field_ulonglong(f, m_row.stat[f->field_index()]);
+    }  // if
+  }    // for
+
+  return 0;
+}
+
+THR_LOCK table_malloc_stats::m_table_lock;
+
+static const char *row_type[] = {"small", "large", "huge"};
+#define NUM_ROWS (sizeof(row_type) / sizeof(row_type[0]))
+static const char *stat_type[NUM_STAT] = {"allocated", "nmalloc", "ndmalloc",
+                                          "nrequests"};
+
+Plugin_table table_malloc_stats::m_table_def(
+    /* Schema name */
+    "performance_schema",
+    /* Name */
+    "malloc_stats",
+    /* Definition */
+    " TYPE char(8) not null,\n"
+    " ALLOCATED bigint(20) unsigned not null,\n"
+    " NMALLOC bigint(20) unsigned not null,\n"
+    " NDALLOC bigint(20) unsigned not null,\n"
+    " NREQUESTS bigint(20) unsigned not null\n",
+    /* Options */
+    " ENGINE=PERFORMANCE_SCHEMA",
+    /* Tablespace */
+    nullptr);
+
+PFS_engine_table_share table_malloc_stats::m_share = {
+    &pfs_readonly_acl,
+    table_malloc_stats::create,
+    nullptr, /* write_row */
+    nullptr, /* delete_all_rows */
+    table_malloc_stats::get_row_count,
+    sizeof(PFS_simple_index),
+    &m_table_lock,
+    &m_table_def,
+    false, /* perpetual */
+    PFS_engine_table_proxy(),
+    {0},
+    false /* m_in_purgatory */
+};
+
+PFS_engine_table *table_malloc_stats::create(PFS_engine_table_share *) {
+  return new table_malloc_stats();
+}
+
+ha_rows table_malloc_stats::get_row_count() { return 5; }
+
+table_malloc_stats::table_malloc_stats()
+    : PFS_engine_table(&m_share, &m_pos),
+      m_row_exists(false),
+      m_pos(0),
+      m_next_pos(0) {}
+
+void table_malloc_stats::reset_position(void) {
+  m_pos.m_index = 0;
+  m_next_pos.m_index = 0;
+}
+
+int table_malloc_stats::rnd_next(void) {
+  m_pos.set_at(&m_next_pos);
+
+  if (m_pos.m_index < NUM_ROWS) {
+    make_row(m_pos.m_index);
+    m_next_pos.set_after(&m_pos);
+    return 0;
+  }
+
+  return HA_ERR_END_OF_FILE;
+}
+
+int table_malloc_stats::rnd_pos(const void *pos) {
+  set_position(pos);
+
+  if (m_pos.m_index < NUM_ROWS) {
+    make_row(m_pos.m_index);
+    return 0;
+  }
+
+  return HA_ERR_RECORD_DELETED;
+}
+
+void table_malloc_stats::make_row(int index) {
+  char s[64];
+  size_t sz = sizeof(epoch);
+
+  memset(&m_row, 0, sizeof(m_row));
+
+  epoch++;
+  jemalloc_mallctl("epoch", nullptr, 0, &epoch, sz);
+
+  m_row.type = index;
+
+  unsigned n = 0;
+  sz = sizeof(unsigned);
+  /* Reading the index of merged arenas statistics, which equals to
+   * number of arenas. */
+  jemalloc_mallctl("arenas.narenas", &n, &sz, nullptr, 0);
+
+  sz = sizeof(size_t);
+  for (unsigned i = 0; i < NUM_STAT; i++) {
+    sprintf(s, "stats.arenas.%u.%s.%s", n, row_type[m_row.type], stat_type[i]);
+    jemalloc_mallctl(s, &m_row.stat[i], &sz, nullptr, 0);
```

Comment:
What happens if jemalloc_mallctl fails here (will not grab stat). Then m_row_stat[i] is uninitialized and we will see garbage in the table?

Reply:
Unfortunately that's so. I will modify the code so we set it to 0 in case of an error.

Reply:
I have just noticed that we were already doing that, at the beginning of the method we zero the array. I will revert these changes.